### PR TITLE
Add officehome for Deep models

### PR DESCRIPTION
**Pull request by YanisLalou**
Old OfficeHome with preprocessing, now called: `OfficeHomeResnet`
OfficeHome for deep models, now called: `OfficeHome`

### CODE REVIEW

File: datasets/officehome.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    from sklearn.preprocessing import LabelEncoder
+    from skada.utils import source_target_merge
+    import torchvision.transforms as transforms
+    from torch.utils.data import DataLoader
+    from PIL import Image
+    from pathlib import Path
+    import sys
+
+    PATH_skada_bench = Path(__file__).resolve().parents[1]
+    sys.path.extend([str(PATH_skada_bench)])
```
Comment (antoinecollas) [inline datasets/officehome.py:13]: looks ugly...

File: benchmark_utils/backbones_architecture.py [hunk 2]
```diff
@@ -33,6 +34,25 @@ def forward(self, x, sample_weight=None):
 
         return x
 
+class OfficeConvNet(nn.Module):
```
Comment (antoinecollas) [inline benchmark_utils/backbones_architecture.py:37]: this neural network is not specific to the office dataset, right? can you call it `ResNet50` ?

File: benchmark_utils/backbones_architecture.py [hunk 3]
```diff
@@ -33,6 +34,25 @@ def forward(self, x, sample_weight=None):
...
+        num_ftrs = self.resnet.fc.in_features
+        self.fc1 = nn.Linear(num_ftrs, 1024)
+        self.fc2 = nn.Linear(1024, n_classes)
+        
+        # Replace ResNet's fc layer with Identity
+        self.resnet.fc = nn.Identity()
+
+    def forward(self, x, sample_weight=None):
+        x = self.resnet(x)
+        x = F.relu(self.fc1(x))
```
Comment (antoinecollas) [inline benchmark_utils/backbones_architecture.py:53]: In the original Resnet50, I think there is on a fully connected layer.

File: benchmark_utils/backbones_architecture.py [hunk 4]
```diff
@@ -33,6 +34,25 @@ def forward(self, x, sample_weight=None):
...
+        x = self.fc2(x)
+        return x
```
Comment (antoinecollas) [inline benchmark_utils/backbones_architecture.py:55]: An overall comment: did you see a paper applying this architecture to Officehome? Maybe a Resnet18 is enough?

File: benchmark_utils/utils.py [hunk 5]
```diff
@@ -17,7 +17,7 @@ def _download_file_with_progress(url, filename):
     # is sometimes so long that it seems broken
     response = requests.get(url, stream=True, timeout=5)
     total_size_in_bytes = int(response.headers.get('content-length', 0))
-    block_size = 16384 #16 Kilobyte
+    block_size = 16384  # 16 Kilobyte
```
Comment (antoinecollas) [inline benchmark_utils/utils.py:20]: are you using a linter that formats code?

File: benchmark_utils/utils.py [hunk 6]
```diff
@@ -97,5 +99,38 @@ def __getitem__(self, idx):
...
         if self.transform:
             image = self.transform(image)
-        
-        return image, label
\ No newline at end of file
+
+        return image, label
+
+
+def get_model_batchsize_for_dataset(dataset_name, n_classes):
```
Comment (antoinecollas) [inline benchmark_utils/utils.py:106]: I suggest `get_model_and_batch_size`

Comment (antoinecollas) [reply benchmark_utils/utils.py:106]: minor comment, you could do something like this:

```python
def get_model_and_batch_size(dataset_name, n_classes):
    """
    Get the appropriate model and batch size for a given dataset.

    Args:
        dataset_name (str): Name of the dataset ('mnist_usps', 'office31', 'officehome', or 'bci').
        n_classes (int): Number of classes in the dataset.

    Returns:
        tuple: A tuple containing:
            - model: The appropriate neural network model for the dataset.
            - batch_size (int): The recommended batch size for the dataset.

    Raises:
        ValueError: If an unsupported dataset name is provided.
    """
    
    dataset_configs = {
        'mnist_usps': {'batch_size': 256, 'model': ShallowConvNet(n_classes=n_classes)},
        'office31': {'batch_size': 128, 'model': OfficeConvNet(n_classes=n_classes)},
        'officehome': {'batch_size': 128, 'model': OfficeConvNet(n_classes=n_classes)},
        'bci': {'batch_size': 256, 'model': ShallowMLP(input_dim=253, n_classes=n_classes).double()}
    }
    
    if dataset_name not in dataset_configs:
        raise ValueError(f"Unsupported dataset: {dataset_name}")
    
    config = dataset_configs[dataset_name]
    return config['model'], config['batch_size']
```

File: solvers/deep_coral.py [hunk 7]
```diff
@@ -5,7 +5,8 @@
 # - getting requirements info when all dependencies are not installed.
 with safe_import_context() as import_ctx:
     from benchmark_utils.base_solver import DASolver
-    from benchmark_utils.backbones_architecture import ShallowConvNet, ShallowMLP
+    from benchmark_utils.backbones_architecture import ShallowConvNet, ShallowMLP, OfficeConvNet
```
Comment (antoinecollas) [inline solvers/deep_coral.py:8]: you can remove this import

File: datasets/officehome.py [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        # Define transformations to preprocess the images
+        preprocess = transforms.Compose([
+            transforms.Resize(256),
+            transforms.CenterCrop(224),
+            transforms.ToTensor(),
+            transforms.Normalize(
+                mean=[0.485, 0.456, 0.406],
+                std=[0.229, 0.224, 0.225],
+            ),
+        ])
```
Comment (antoinecollas) [inline datasets/officehome.py:51]: where did you get these values?

File: datasets/officehome.py [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+        # Create a DataLoader for the dataset
+        dataset = ImageDataset(
+            self.path_extract, transform=preprocess, domain_select=domain_select
+        )
```
Comment (antoinecollas) [inline datasets/officehome.py:56]: This is unrelated to this PR, but wouldn't it be better to integrate this ImageDataset into Skada?

File: datasets/officehome.py [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+        le = LabelEncoder()
+        le.fit(np.concatenate([y_source, y_target]))
+        y_source = le.transform(y_source)
+        y_target = le.transform(y_target)
+
+        X, y, sample_domain = source_target_merge(
+            X_source, X_target, y_source, y_target)
+
+        print(f'OfficeHome mean {X.mean()}')
+        print(f'OfficeHome std {X.std()}')
```
Comment (antoinecollas) [inline datasets/officehome.py:91]: great, do you get 0 and 1?

File: datasets/officehome.py [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+
+        # Create a DataLoader for the dataset
+        dataset = ImageDataset(
+            self.path_extract, transform=preprocess, domain_select=domain_select
+        )
+        dataloader = DataLoader(
+            dataset, batch_size=len(dataset), shuffle=False
+        )
+
+        images, labels = next(iter(dataloader))
```
Comment (antoinecollas) [inline datasets/officehome.py:61]: Why do you need to do this?

File: solvers/deep_coral.py [hunk 12]
```diff
@@ -37,30 +37,21 @@ def get_estimator(self, n_classes, device, dataset_name, **kwargs):
...
-        
+        model, batch_size = get_model_batchsize_for_dataset(
+            dataset_name, n_classes)
+
         lr_scheduler = LRScheduler(
             policy='StepLR',
             step_every='epoch',
             step_size=1,
             gamma=0.7
         )
```
Comment (antoinecollas) [inline solvers/deep_coral.py:48]: where did you get this?

**Review by antoinecollas — COMMENTED** [review]
I wrote a few comments. Otherwise, the PR looks great!

File: datasets/officehome.py (hunk 9 above)
Comment (tgnassou) [reply datasets/officehome.py:56]: We don't want a lot of dataset in skada

File: solvers/deep_coral.py (hunk 12 above)
Comment (tgnassou) [reply solvers/deep_coral.py:48]: It's in the torch example for MNIST, maybe not relevant for all datasets

File: datasets/officehome.py (hunk 9 above)
Comment (antoinecollas) [reply datasets/officehome.py:56]: but `ImageDataset` is just a wrapper to get, preprocess, and select a domain, right?

File: datasets/officehome.py (hunk 8 above)
Comment (antoinecollas) [reply datasets/officehome.py:51]: The values are the ones used to train a resnet on imagenet. Then, from https://pytorch.org/tutorials/beginner/transfer_learning_tutorial.html we should use these values for transfer learning.

File: datasets/officehome.py (hunk 10 above)
Comment (antoinecollas) [reply datasets/officehome.py:91]: a priori, no since the normalization from ImageNet is used

File: benchmark_utils/utils.py (hunk 5 above)
Comment (YanisLalou) [reply benchmark_utils/utils.py:20]: yes, the default one in vs code I assume?

**Pull request merged**